### DISCUSSION TITLE
custom firmware build error

### DISCUSSION THREAD

**Discussion by gitrimius** [shanteacontrols/OpenDeck#38] [Q&A]
Hi,
I am trying to compile custom firmware for STM32F411 and 1x4051 muxes as a test on Ubuntu20.
I get following error when the building stops. Any idea what is going wrong?
This is the error (i used the standard yml file for 411 to drive a test...):
Thanks a lot!

> 
> Building application...
> Generating pins...
> Building: board/gen/blackpill411/Pins.cpp
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ctlreq.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_core.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ioreq.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_spi.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_exti.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ramfunc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.c
> Building: board/stm32/gen/f4/common/Src/usbd_conf.c
> Building: board/stm32/gen/f4/common/Src/system_stm32f4xx.c
> Building: board/stm32/variants/f4/Common.cpp
> Building: board/stm32/variants/f4/Flash.cpp
> Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
> Building: board/common/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/startup.s
> Building: board/stm32/variants/f4/stm32f411ce/FlashPages.cpp
> Building: board/stm32/common/Multiplexer.cpp
> Building: board/stm32/common/ISR.cpp
> Building: board/stm32/common/NVM.cpp
> Building: board/stm32/common/Init.cpp
> Building: board/stm32/common/Bootloader.cpp
> Building: board/stm32/common/ShiftRegistersWait.cpp
> Building: common/OpenDeckMIDIformat/OpenDeckMIDIformat.cpp
> Building: board/stm32/usb/midi/USB.cpp
> In file included from board/stm32/usb/midi/USB.cpp:20:
> In file included from ./board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Inc/usbd_core.h:29:
> In file included from ./board/stm32/gen/f4/common/Inc/usbd_conf.h:24:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal.h:30:
> In file included from board/stm32/variants/f4/stm32f4xx_hal_conf.h:375:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_pcd.h:29:
> <b>./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_ll_usb.h:340:9: </b><font color="#A347BA"><b>warning: </b></font><b>&apos;EP_TYPE_BULK&apos; macro redefined [-Wmacro-redefined]</b>
> #define EP_TYPE_BULK                           2U
> <font color="#26A269"><b>        ^</b></font>
> <b>./board/common/usb/descriptors/types/Helpers.h:103:9: </b><font color="#171421"><b>note: </b></font>previous definition is here
> #define EP_TYPE_BULK 0x02
> <font color="#26A269"><b>        ^</b></font>
> 1 warning generated.
> Building: board/common/usb/descriptors/midi/Descriptors.c
> Building: board/stm32/uart/UART.cpp
> Building: board/common/uart/UART.cpp
> Building: application/main.cpp
> Building: application/database/CustomInit.cpp
> Building: application/database/Database.cpp
> Building: application/system/Get.cpp
> Building: application/system/System.cpp
> Building: application/system/Set.cpp
> Building: application/io/common/Common.cpp
> Building: ../modules/sysex/src/SysExConf.cpp
> Building: ../modules/midi/src/MIDI.cpp
> Building: ../modules/dbms/src/LESSDB.cpp
> Building: board/common/io/Stubs.cpp
> Building: application/io/analog/Analog.cpp
> Building: board/common/io/Analog.cpp
> Building: application/io/leds/LEDs.cpp
> Building: board/common/io/Output.cpp
> Building: application/io/buttons/Buttons.cpp
> Building: application/io/encoders/Encoders.cpp
> Building: board/common/io/Input.cpp
> Building: application/io/touchscreen/Touchscreen.cpp
> Building: application/io/touchscreen/model/viewtech/Viewtech.cpp
> Building: application/io/touchscreen/model/nextion/Nextion.cpp
> Building: application/io/display/strings/Strings.cpp
> Building: application/io/display/Display.cpp
> Building: application/io/display/U8X8/U8X8.cpp
> Building: board/stm32/i2c/I2C.cpp
> Building: ../modules/u8g2/csrc/u8x8_string.c
> Building: ../modules/u8g2/csrc/u8x8_setup.c
> Building: ../modules/u8g2/csrc/u8x8_u8toa.c
> Building: ../modules/u8g2/csrc/u8x8_8x8.c
> Building: ../modules/u8g2/csrc/u8x8_u16toa.c
> Building: ../modules/u8g2/csrc/u8x8_display.c
> Building: ../modules/u8g2/csrc/u8x8_fonts.c
> Building: ../modules/u8g2/csrc/u8x8_byte.c
> Building: ../modules/u8g2/csrc/u8x8_cad.c
> Building: ../modules/u8g2/csrc/u8x8_gpio.c
> Building: ../modules/u8g2/csrc/u8x8_d_ssd1306_128x64_noname.c
> Building: ../modules/u8g2/csrc/u8x8_d_ssd1306_128x32.c
> Building: application/database/CustomInit.cpp
> Building: application/database/Database.cpp
> Building: ../modules/dbms/src/LESSDB.cpp
> Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
> Building: board/stm32/variants/f4/stm32f411ce/FlashPages.cpp
> Building: flashgen/main.cpp
> Creating executable: build/flashgen/blackpill411/release/blackpill411.elf
> Generating factory flash configuration...
> Creating executable: build/app/blackpill411/release/blackpill411.elf
>    text	   data	    bss	    dec	    hex	filename
>   69724	     24	   7364	  77112	  12d38	./build/app/blackpill411/release/blackpill411.elf
> Building bootloader...
> Building: board/gen/blackpill411/Pins.cpp
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ctlreq.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_core.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ioreq.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_spi.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_exti.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ramfunc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.c
> Building: board/stm32/gen/f4/common/Src/usbd_conf.c
> Building: board/stm32/gen/f4/common/Src/system_stm32f4xx.c
> Building: board/stm32/variants/f4/Common.cpp
> Building: board/stm32/variants/f4/Flash.cpp
> Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
> Building: board/common/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/startup.s
> Building: board/stm32/variants/f4/stm32f411ce/FlashPages.cpp
> Building: board/common/bootloader/Bootloader.cpp
> Building: board/common/io/Indicators.cpp
> Building: board/stm32/common/Bootloader.cpp
> Building: board/stm32/common/Init.cpp
> Building: board/stm32/common/ShiftRegistersWait.cpp
> Building: board/stm32/common/ISR.cpp
> Building: bootloader/SysExParser/SysExParser.cpp
> Building: bootloader/main.cpp
> Building: bootloader/updater/Updater.cpp
> Building: board/common/usb/descriptors/midi/Descriptors.c
> Building: board/stm32/usb/midi/USB.cpp
> In file included from board/stm32/usb/midi/USB.cpp:20:
> In file included from ./board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Inc/usbd_core.h:29:
> In file included from ./board/stm32/gen/f4/common/Inc/usbd_conf.h:24:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal.h:30:
> In file included from board/stm32/variants/f4/stm32f4xx_hal_conf.h:375:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_pcd.h:29:
> <b>./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_ll_usb.h:340:9: </b><font color="#A347BA"><b>warning: </b></font><b>&apos;EP_TYPE_BULK&apos; macro redefined [-Wmacro-redefined]</b>
> #define EP_TYPE_BULK                           2U
> <font color="#26A269"><b>        ^</b></font>
> <b>./board/common/usb/descriptors/types/Helpers.h:103:9: </b><font color="#171421"><b>note: </b></font>previous definition is here
> #define EP_TYPE_BULK 0x02
> <font color="#26A269"><b>        ^</b></font>
> 1 warning generated.
> Creating executable: build/boot/blackpill411/release/blackpill411.elf
>    text	   data	    bss	    dec	    hex	filename
>   25212	     16	   3940	  29168	   71f0	./build/boot/blackpill411/release/blackpill411.elf
> Building CDC...
> Building: board/gen/blackpill411/Pins.cpp
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ctlreq.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_core.c
> Building: board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ioreq.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_spi.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_exti.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ramfunc.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd.c
> Building: board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.c
> Building: board/stm32/gen/f4/common/Src/usbd_conf.c
> Building: board/stm32/gen/f4/common/Src/system_stm32f4xx.c
> Building: board/stm32/variants/f4/Common.cpp
> Building: board/stm32/variants/f4/Flash.cpp
> Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
> Building: board/common/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/Common.cpp
> Building: board/stm32/variants/f4/stm32f411ce/startup.s
> Building: board/stm32/variants/f4/stm32f411ce/FlashPages.cpp
> Building: board/stm32/common/Init.cpp
> Building: board/common/io/Indicators.cpp
> Building: board/stm32/common/ShiftRegistersWait.cpp
> Building: board/stm32/common/ISR.cpp
> In file included from board/stm32/common/ISR.cpp:24:
> In file included from ./board/common/usb/descriptors/cdc/Descriptors.h:51:
> In file included from ./board/common/usb/Arch.h:28:
> In file included from ./board/common/usb/descriptors/types/CDCDescriptors.h:43:
> In file included from ./board/common/usb/descriptors/types/StdDescriptors.h:44:
> <b>./board/common/usb/descriptors/types/Helpers.h:103:9: </b><font color="#A347BA"><b>warning: </b></font><b>&apos;EP_TYPE_BULK&apos; macro redefined [-Wmacro-redefined]</b>
> #define EP_TYPE_BULK 0x02
> <font color="#26A269"><b>        ^</b></font>
> <b>./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_ll_usb.h:340:9: </b><font color="#171421"><b>note: </b></font>previous definition is here
> #define EP_TYPE_BULK                           2U
> <font color="#26A269"><b>        ^</b></font>
> 1 warning generated.
> Building: board/stm32/uart/UART.cpp
> Building: board/common/uart/UART.cpp
> Building: board/common/usb/descriptors/cdc/Descriptors.c
> Building: board/stm32/usb/cdc/USB.cpp
> In file included from board/stm32/usb/cdc/USB.cpp:20:
> In file included from ./board/stm32/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Inc/usbd_core.h:29:
> In file included from ./board/stm32/gen/f4/common/Inc/usbd_conf.h:24:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal.h:30:
> In file included from board/stm32/variants/f4/stm32f4xx_hal_conf.h:375:
> In file included from ./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_pcd.h:29:
> <b>./board/stm32/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_ll_usb.h:340:9: </b><font color="#A347BA"><b>warning: </b></font><b>&apos;EP_TYPE_BULK&apos; macro redefined [-Wmacro-redefined]</b>
> #define EP_TYPE_BULK                           2U
> <font color="#26A269"><b>        ^</b></font>
> <b>./board/common/usb/descriptors/types/Helpers.h:103:9: </b><font color="#171421"><b>note: </b></font>previous definition is here
> #define EP_TYPE_BULK 0x02
> <font color="#26A269"><b>        ^</b></font>
> 1 warning generated.
> Building: cdc/main.cpp
> Creating executable: build/cdc/blackpill411/release/blackpill411.elf
>    text	   data	    bss	    dec	    hex	filename
>   20504	    228	  19388	  40120	   9cb8	./build/cdc/blackpill411/release/blackpill411.elf
> Merged binary created: ./build/merged/blackpill411.hex
> </pre>

**Comment by paradajz** [comment]
Those are not errors, only warnings. Everything is fine,merged binary is created.

-------- Original Message --------

**Reply by gitrimius** [reply]
okay! thanks a lot!
I would like to use the occasion to congratulate you with this project! Very nice and user-friendliness is great in that regard that a noob like me was able to install the main fw, play with the settings, make a custom fw without hurting his brains and room walls..
Really nice!
Is there a way I can support your project with a small financial transaction. I think that is more than appropriate!

**Reply by paradajz** [reply]
Thank you. I'm glad you've found the project useful. You can mail me if you really want to make a donation.
I've just pushed fixes for those warnings, by the way.
